File: Android /Android/25:Jan:2017/FishbowlMobileApps/AndroidSDK/FishbowlMobileApps-Jamba-Android-v2.4.2/JMBA_Olo_iOS_SDK/Olo/OloSessionService.swift
//
//  OloSessionService.swift
//  Olo
//
//  Created by Taha Samad on 4/28/15.
//  Copyright (c) 2015 Hathway, Inc. All rights reserved.
//

import Foundation
import SwiftyJSON

public typealias OloSessionUserCallback = (_ user: OloUser?, _ error: NSError?) -> Void
public typealias OloSessionAuthTokenCallback = (_ oloAuthToken: String?, _ error: NSError?) -> Void
public typealias OloSessionInvalidAuthTokenCallback = (_ request:URLRequest , _ res:HTTPURLResponse, _ jsonObject:JSON, _ callback: OloServiceCallback) -> Void

open class OloSessionService {
    
    open static var authToken: String?
    
    //This callback if set can be used to intercept and provide centralized handling for Invalid Auth Error Olo Error Code 100,101
    //It is responsibility of the this callback to call the passed callback with appropiate parameter
    open static var invalidAuthTokenCallback: OloSessionInvalidAuthTokenCallback?
    
    open class func createUser(_ firstName: String, lastName: String, contactNumber: String, emailAddress: String, password: String, callback: @escaping OloSessionUserCallback) {
        let payload = [
            "firstname": firstName,
            "lastname": lastName,
            "emailaddress": emailAddress,
            "contactnumber": contactNumber,
            "password": password
        ]
        OloService.post("/users/create", parameters: payload as OloJSONDictionary) { (response, error) -> Void in
            if error != nil {
                callback(nil, error)
                return
            }
            let user = OloUser(json: response)
            self.authToken = user.authToken
            callback(user, nil)
        }
    }
    
    open class func authenticateUser(_ login: String, password: String, basketId: String?, callback: @escaping OloSessionUserCallback) {
        var payload = [
            "login": login,
            "password": password
        ]
        if basketId != nil {
            payload["basketid"] = basketId!
        }
        OloService.post("/users/authenticate", parameters: payload) { (response, error) -> Void in
            if error != nil {
                callback(nil, error)
                return
            }
            let user = OloUser(json: response["user"])
            self.authToken = user.authToken
            callback(user, nil)
        }
    }
    
    // If user is authenticated, remove token from the API, but return immediately
    // Ignore any possible API errors, as we want to end our user session in any case
    open class func deleteAuthenticationToken() {
        if let authToken = OloSessionService.authToken {
            OloService.delete("/users/\(authToken)", parameters: nil) { (response, error) in
                // Ignore any possible errors
            }
        }
        OloSessionService.authToken = nil
    }

    /// Authenticate or create user with third party vendor
    open class func getOrCreate(_ provider: String, providerToken: String, contactNumber: String? = nil, basketId: String? = nil, callback: @escaping OloSessionAuthTokenCallback) {
        var payload:[String:AnyObject] = [
            "provider": provider as AnyObject,
            "providertoken": providerToken as AnyObject
        ]
        payload["contactnumber"] = contactNumber as AnyObject?
        payload["basketid"] = basketId as AnyObject?
        OloService.post("/users/getorcreate", parameters: payload) { (response, error) -> Void in
            if error != nil {
                callback(nil, error)
                return
            }
            else {
                let authToken = response["authtoken"].stringValue
                self.authToken = authToken
                callback(authToken, nil)
            }
        }
    }

}

File: Android /Android/25:Jan:2017/FishbowlMobileApps/AndroidSDK/FishbowlMobileApps-Jamba-Android-v2.4.2/JMBA_Olo_Android_SDK/app/src/main/java/com/wearehathway/apps/olo/Services/OloSessionService.java
package com.wearehathway.apps.olo.Services;

import com.google.gson.Gson;
import com.wearehathway.apps.olo.Interfaces.OloInvalidAuthCallback;
import com.wearehathway.apps.olo.Interfaces.OloServiceCallback;
import com.wearehathway.apps.olo.Interfaces.OloSessionAuthTokenCallback;
import com.wearehathway.apps.olo.Interfaces.OloSessionServiceCallback;
import com.wearehathway.apps.olo.Interfaces.OloUserLogoutCallback;
import com.wearehathway.apps.olo.Models.OloSession;
import com.wearehathway.apps.olo.Models.OloUser;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;

import static com.wearehathway.apps.olo.Utils.Constants.AUTHENTICATE_USER;
import static com.wearehathway.apps.olo.Utils.Constants.CREATE_USER;

/**
 * Created by Nauman Afzaal on 28/04/15.
 */
public class OloSessionService
{
    public static OloInvalidAuthCallback invalidAuthTokenCallback;

    public static void notifyInValidAuthTokenCalback()
    {
        clearSession();
        if(invalidAuthTokenCallback != null)
        {
            invalidAuthTokenCallback.onInvalidAuthTokenCallback();
        }
    }

    public final static OloSession currentSesstion = new OloSession();

    public static void setSessionParams(String authToken)
    {
        currentSesstion.setAuthToken(authToken);
    }

    public static void clearSession()
    {
        currentSesstion.clearSession();
    }

    public static void createUser(String firstName, String lastName, String contactNumber, String emailAddress, String password, final OloSessionServiceCallback callback)
    {
        HashMap<String, Object> parameters = new HashMap<>();
        parameters.put("firstname", firstName);
        parameters.put("lastname", lastName);
        parameters.put("emailaddress", emailAddress);
        parameters.put("contactnumber", contactNumber);
        parameters.put("password", password);

        OloService.getInstance().post(CREATE_USER, parameters, new OloServiceCallback()
        {
            @Override
            public void onServiceCallback(JSONObject response, Exception error)
            {
                parseResponse(response, error, callback);
            }
        });
    }

    public static void authenticateUser(String email, String password, final OloSessionServiceCallback callback)
    {
        HashMap<String, Object> parameters = new HashMap<>();
        parameters.put("login", email);
        parameters.put("password", password);

        OloService.getInstance().post(AUTHENTICATE_USER, parameters, new OloServiceCallback()
        {
            @Override
            public void onServiceCallback(JSONObject response, Exception error)
            {
                parseResponse(response, error, callback);
            }
        });
    }


    public static void logOutUser(final OloUserLogoutCallback callback)
    {
        String authToken = OloSessionService.currentSesstion.getAuthToken();
        if (authToken == null)
        {
            callback.onUserLogoutCallback(new Exception("User is not authenticated."));
            return;
        }
        String path = "users/" + currentSesstion.getAuthToken();
        OloService.getInstance().delete(path, null, new OloServiceCallback()
        {
            @Override
            public void onServiceCallback(JSONObject response, Exception error)
            {
                if (error == null)
                {
                    currentSesstion.clearSession();
                }
                if (callback != null)
                {
                    callback.onUserLogoutCallback(error);
                }
            }
        });

    }

    public static void getOrCreateUser(String provider, String providerToken, String contactNumber, String basketId, final OloSessionAuthTokenCallback callback)
    {
        String path = "/users/getorcreate";
        HashMap<String, Object> parameters = new HashMap<>();
        parameters.put("provider", provider);
        parameters.put("providertoken", providerToken);
        parameters.put("contactnumber", contactNumber);
        parameters.put("basketid", basketId);
        OloService.getInstance().post(path, parameters, new OloServiceCallback()
        {
            @Override
            public void onServiceCallback(JSONObject response, Exception error)
            {
                String authToken = "";
                try
                {
                    if (response != null)
                    {
                        authToken = response.getString("authtoken");
                        currentSesstion.setAuthToken(authToken);
                    }
                } catch (JSONException e)
                {
                    e.printStackTrace();
                    error = new Exception("Error occurred while parsing data.");
                }
                if (callback != null)
                {
                    callback.onOloSessionAuthTokenCallback(authToken, error);
                }
            }
        });
    }

    private static void parseResponse(JSONObject response, Exception error, OloSessionServiceCallback callback)
    {
        OloUser user = null;
        if (response != null)
        {
            Gson gson = new Gson();
            if (response.has("user"))
            {
                try
                {
                    response = response.getJSONObject("user");
                } catch (JSONException e)
                {
                    e.printStackTrace();
                }
            }
            user = gson.fromJson(response.toString(), OloUser.class);
            currentSesstion.setAuthToken(user.getAuthtoken());
            if (user == null)
            {
                error = new Exception("Error occurred while parsing data.");
            }
        }
        if (callback != null)
        {
            callback.onSessionServiceCallback(user, error);
        }
    }
}
